File: app/src/main/java/me/kirimin/mitsumine/ui/fragment/TagEditDialogFragment.java
package me.kirimin.mitsumine.ui.fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ListView;

import java.util.ArrayList;

import me.kirimin.mitsumine.R;
import me.kirimin.mitsumine.ui.adapter.TagEditDialogFragmentAdapter;
import rx.android.events.OnClickEvent;
import rx.functions.Action1;

public class TagEditDialogFragment extends DialogFragment {

    public interface OnOkClickListener {
        public void onOkClick(ArrayList<String> tags);
    }

    static TagEditDialogFragment newInstance(ArrayList tags, Fragment targetFragment) {
        TagEditDialogFragment fragment = new TagEditDialogFragment();
        Bundle bundle = new Bundle();
        bundle.putStringArrayList("tags", tags);
        fragment.setArguments(bundle);
        fragment.setTargetFragment(targetFragment, 0);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final ArrayList<String> tags = getArguments().getStringArrayList("tags");
        final View rootView = inflater.inflate(R.layout.dialog_fragment_tag_edit, container, false);
        final ListView listView = (ListView) rootView.findViewById(R.id.RegisterBookmarkTagEditDialogListView);
        listView.setAdapter(new TagEditDialogFragmentAdapter(getActivity(), tags, new Action1<OnClickEvent>() {
            @Override
            public void call(OnClickEvent onClickEvent) {
                TagEditDialogFragmentAdapter adapter = (TagEditDialogFragmentAdapter) listView.getAdapter();
                adapter.remove((String) onClickEvent.view.getTag());
            }
        }));
        rootView.findViewById(R.id.RegisterBookmarkTagEditDialogOkButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                TagEditDialogFragmentAdapter adapter = (TagEditDialogFragmentAdapter) listView.getAdapter();
                ((OnOkClickListener) getTargetFragment()).onOkClick(adapter.getAllItem());
                dismiss();
            }
        });
        rootView.findViewById(R.id.RegisterBookmarkTagEditDialogAddButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                EditText tagNameEdit = (EditText) rootView.findViewById(R.id.RegisterBookmarkTagEditDialogAddEditText);
                if (!tagNameEdit.getText().toString().isEmpty()) {
                    TagEditDialogFragmentAdapter adapter = (TagEditDialogFragmentAdapter) listView.getAdapter();
                    adapter.remove(tagNameEdit.getText().toString());
                    adapter.add(tagNameEdit.getText().toString());
                    tagNameEdit.setText("");
                }
            }
        });

        getDialog().setTitle("タグ編集");
        return rootView;
    }
}

File: app/src/main/java/me/kirimin/mitsumine/ui/fragment/TagEditDialogFragment.kt
package me.kirimin.mitsumine.ui.fragment

import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import java.util.ArrayList

import me.kirimin.mitsumine.R
import me.kirimin.mitsumine.ui.adapter.TagEditDialogFragmentAdapter
import rx.functions.Action1

import kotlinx.android.synthetic.dialog_fragment_tag_edit.view.*

public class TagEditDialogFragment : DialogFragment() {

    companion object {

        fun newInstance(tags: ArrayList<String>, targetFragment: Fragment): TagEditDialogFragment {
            val fragment = TagEditDialogFragment()
            val bundle = Bundle()
            bundle.putStringArrayList("tags", tags)
            fragment.arguments = bundle
            fragment.setTargetFragment(targetFragment, 0)
            return fragment
        }
    }

    public interface OnOkClickListener {
        public fun onOkClick(tags: ArrayList<String>)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val tags = arguments.getStringArrayList("tags")
        val rootView = inflater.inflate(R.layout.dialog_fragment_tag_edit, container, false)
        rootView.listView.adapter = TagEditDialogFragmentAdapter(activity, tags, Action1 { onClickEvent ->
            val adapter = rootView.listView.adapter as TagEditDialogFragmentAdapter
            adapter.remove(onClickEvent.view().tag as String)
        })
        rootView.okButton.setOnClickListener {
            val adapter = rootView.listView.adapter as TagEditDialogFragmentAdapter
            (targetFragment as OnOkClickListener).onOkClick(adapter.getAllItem())
            dismiss()
        }
        rootView.addButton.setOnClickListener {
            if (!rootView.tagNameEditText.text.toString().isEmpty()) {
                val adapter = rootView.listView.adapter as TagEditDialogFragmentAdapter
                adapter.remove(rootView.tagNameEditText.text.toString())
                adapter.add(rootView.tagNameEditText.text.toString())
                rootView.tagNameEditText.setText("")
            }
        }

        getDialog().setTitle(R.string.register_bookmark_tag_edit)
        return rootView
    }
}
